services/galley/src/Galley/API/Teams.hs

```diff
@@ -493,10 +495,22 @@ getTeamMembers ::
   Maybe TeamMembersPagingState ->
   Sem r TeamMembersPage
 getTeamMembers lzusr tid mbMaxResults mbPagingState = do
-  member <- E.getTeamMember tid (tUnqualified lzusr) >>= noteS @'NotATeamMember
+  let uid = tUnqualified lzusr
+  member <- E.getTeamMember tid uid >>= noteS @'NotATeamMember
   let mState = C.PagingState . LBS.fromStrict <$> (mbPagingState >>= mtpsState)
   let mLimit = fromMaybe (unsafeRange Public.hardTruncationLimit) mbMaxResults
-  E.listTeamMembers @CassandraPaging tid mState mLimit <&> toTeamMembersPage member
+  if member `hasPermission` SearchContacts
+    then E.listTeamMembers @CassandraPaging tid mState mLimit <&> toTeamMembersPage member
+    else do
+      -- If the user does not have the SearchContacts permission (e.g. the external partner),
+      -- we only return the team members they are in a conversation with plus the invitee
+      teamCids <- fmap (\tConv -> tConv ^. conversationId) <$> E.getTeamConversations tid
+      userTeamCids <- E.selectConversations uid teamCids
+      convs <- E.getConversations userTeamCids
+      let lConvMems = Conv.lmId <$> (convs >>= convLocalMembers)
+      let invitee = member ^. invitation <&> fst
+      let uids = nub (maybe (uid : lConvMems) (: uid : lConvMems) invitee)
+      E.selectTeamMembersPaginated tid uids mState mLimit <&> toTeamMembersPage member
```

Comment:
This is not very safe to do for large teams. I think we should create another index for these things.